### PR TITLE
Add bytecode compiler and runner

### PR DESCRIPTION
**Pull request by mohanson**

### CODE REVIEW

File: quickjs/qjs.c [hunk 1]
```diff
@@ -95,6 +97,43 @@ static int eval_buf(JSContext *ctx, const void *buf, int buf_len, const char *fi
...
+    val = JS_EvalFunction(ctx, obj);
+    if (JS_IsException(val)) {
+        js_std_dump_error(ctx);
+        return -1;
+    }
+    JS_FreeValue(ctx, val);
+    return 0;
+}
+
+static int compile(JSContext *ctx, const void *buf, int buf_len) {
```
Comment (XuJiandong) [inline quickjs/qjs.c:118]: `compile` feature is not necessary. We can use native `qjsc`.

File: quickjs/qjs.c [hunk 2]
```diff
@@ -125,12 +164,20 @@ static JSContext *JS_NewCustomContext(JSRuntime *rt) {
...
 static bool s_local_access = false;
+static bool s_compile = false;
+static bool s_bytecode = false;
 
 static void parse_args(int argc, const char **argv) {
     for (int i = 0; i < argc; i++) {
         if (strcmp(argv[i], "-r") == 0) {
             s_local_access = true;
         }
+        if (strcmp(argv[i], "-b") == 0) {
```
Comment (XuJiandong) [inline quickjs/qjs.c:175]: Try to guess byte code without flag.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: quickjs/qjs.c [hunk 3]
```diff
@@ -114,11 +128,102 @@ void js_std_dump_error(JSContext *ctx) {
...
+exit:
+    return err;
+}
+
 static int eval_buf(JSContext *ctx, const void *buf, int buf_len, const char *filename, int eval_flags) {
     JSValue val;
     int ret;
 
-    if ((eval_flags & JS_EVAL_TYPE_MASK) == JS_EVAL_TYPE_MODULE) {
+    if (((const char *)buf)[0] == (char)0x02) {
```
Comment (XuJiandong) [inline quickjs/qjs.c:211]: `0x02` can be defined as a macro.

File: tests/ckb_js_tests/Makefile [hunk 4]
```diff
@@ -19,7 +19,14 @@ build/testdata_fs_modules.bin: test_data/fs_module/main.js test_data/fs_module/f
...
 
 fs_def_cell_module: build/testdata_fs_modules.bin
-	$(CKB_DEBUGGER) --max-cycles $(MAX_CYCLES) ---read-file ../../$? --bin $(CKB_JS_VM_BIN_PATH) -- -f -r  2>&1 | fgrep 'Run result: 0'
+	$(CKB_DEBUGGER) --max-cycles $(MAX_CYCLES) --read-file ../../$? --bin $(CKB_JS_VM_BIN_PATH) -- -f -r  2>&1 | fgrep 'Run result: 0'
+
+fs_def_cell_module_bytecode: build/testdata_fs_modules.bin
+	mkdir -p ../../build/bytecode
+	$(CKB_DEBUGGER) --read-file ../../build/testdata_fs_modules.bin --bin $(CKB_JS_VM_BIN_PATH) -- -c -f main.js | head -n -3 | xxd -r -p > ../../build/bytecode/main.js
+	$(CKB_DEBUGGER) --read-file ../../build/testdata_fs_modules.bin --bin $(CKB_JS_VM_BIN_PATH) -- -c -f fib_module.js | head -n -3 | xxd -r -p > ../../build/bytecode/fib_module.js
+	cd ../../build/bytecode && lua ../../tools/fs.lua pack ../../build/testdata_fs_modules_bc.bin main.js fib_module.js
```
Comment (XuJiandong) [inline tests/ckb_js_tests/Makefile:28]: rename "*.js" to "*.bc".

File: quickjs/qjs.c [hunk 5]
```diff
@@ -114,11 +128,102 @@ void js_std_dump_error(JSContext *ctx) {
...
+    for (int i = 0; i < out_buf_len; i += 32) {
+        uint32_t size = i + 32 > out_buf_len ? out_buf_len - i : 32;
+        _exec_bin2hex(&out_buf[i], size, msg_buf, 65, &size, true);
+        msg_buf[size - 1] = 0;
+        printf("%s", msg_buf);
+    }
+    return 0;
+}
+
+int compile_from_file_system(JSContext *ctx, const char* filename) {
```
Comment (XuJiandong) [inline quickjs/qjs.c:164]: I think "-c -f" options is not necessary. "-c" option is enough.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: quickjs/qjs.c (hunk 5 above)
Comment (XuJiandong) [reply quickjs/qjs.c:164]: It can also be achieved by "-c". Compiling a file in file system is not necessary.

Comment (mohanson) [reply quickjs/qjs.c:164]: If the javascript code imports a module (`import { fib } from "./fib_module.js"`, eg), we cannot compile it without telling it the contents of `fib_module.js`, therefore a file system is necessary

**Review by XuJiandong — APPROVED** [review]

**Pull request merged**